src/sparse.jl

```diff
@@ -0,0 +1,129 @@
+struct SparseValueIterationSolver <: Solver
+    max_iterations::Int64
+    belres::Float64 # the Bellman Residual
+    include_Q::Bool
+    verbose::Bool
+    init_util::Vector{Float64}
+end
+
+function SparseValueIterationSolver(;max_iterations=500,
+                                    belres::Float64=1e-3,
+                                    include_Q::Bool=true,
+                                    verbose::Bool=false,
+                                    init_util::Vector{Float64}=Vector{Float64}(undef, 0))
+    return SparseValueIterationSolver(max_iterations, belres, include_Q, verbose, init_util)
+end
+
+@POMDP_require solve(solver::SparseValueIterationSolver, mdp::Union{MDP,POMDP}) begin
+    P = typeof(mdp)
+    S = statetype(P)
+    A = actiontype(P)
+    @req discount(::P)
+    @req n_states(::P)
+    @req n_actions(::P)
+    @subreq ordered_states(mdp)
+    @subreq ordered_actions(mdp)
+    @req transition(::P,::S,::A)
+    @req reward(::P,::S,::A,::S)
+    @req stateindex(::P,::S)
+    @req actionindex(::P, ::A)
+    @req actions(::P, ::S)
+    as = actions(mdp)
+    ss = states(mdp)
+    a = first(as)
+    s = first(ss)
+    dist = transition(mdp, s, a)
+    D = typeof(dist)
+    @req support(::D)
+    @req pdf(::D,::S)
+end
+
+function qvalue!(m::Union{MDP,POMDP}, transition_A_S_S2, reward_S::AbstractVector{F}, value_S::AbstractVector{F}, out_qvals_S_A) where {F}
+    @assert size(out_qvals_S_A) == (n_states(m), n_actions(m))
+    for a in 1:n_actions(m)
+        out_qvals_S_A[:, a] = reward_S + discount(m) * transition_A_S_S2[a] * value_S
+    end
+end
+
+function transition_matrix_a_s_sp(mdp::MDP)
+    # Thanks to zach
+    na = n_actions(mdp)
+    ns = n_states(mdp)
+    transmat_row_A = [Float64[] for _ in 1:n_actions(mdp)]
+    transmat_col_A = [Float64[] for _ in 1:n_actions(mdp)]
+    transmat_data_A = [Float64[] for _ in 1:n_actions(mdp)]
+
+    for a in actions(mdp)
+        ai = actionindex(mdp, a)
+        for s in states(mdp)
+            si = stateindex(mdp, s)
+            if !isterminal(mdp, s) # if terminal, the transition probabilities are all just zero
+                td = transition(mdp, s, a)
+                for (sp, p) in weighted_iterator(td)
+                    if p > 0.0
+                        spi = stateindex(mdp, sp)
+                        push!(transmat_row_A[ai], si)
+                        push!(transmat_col_A[ai], spi)
+                        push!(transmat_data_A[ai], p)
+                    end
+                end
+            else
+                push!(transmat_row_A[ai], si)
+                push!(transmat_col_A[ai], si)
+                push!(transmat_data_A[ai], 1.0)
```

Comment:
I think this should be 0.0 right (i.e. unassigned in the sparse matrix)? What if there is a terminal reward, e.g. `reward(m, s, a, sp) = isterminal(m, sp)`, then it will keep getting that reward over and over again?

Reply:
0.0 doesn't work. Transition probabilities don't sum to 1.

I added a note:
```
                # Terminal state
                # Note: Assumes that reward is R(s,a) and sp is absorbing state
```
Otherwise it won't be as efficient.